test/md5.cpp

```diff
@@ -44,76 +39,46 @@ namespace boost {
 
 BOOST_TEST_DONT_PRINT_LOG_VALUE(md5::construction::type::digest_type)
 
-static const std::unordered_map<std::string, std::string> string_data = {
-    {"", "d41d8cd98f00b204e9800998ecf8427e"},
-    {"a", "0cc175b9c0f1b6a831c399e269772661"},
-    {"abc", "900150983cd24fb0d6963f7d28e17f72"},
-    {"message digest", "f96b697d7cb7938d525a2f31aaf161d0"},
-    {"abcdefghijklmnopqrstuvwxyz", "c3fcd3d76192e4007dfb496cca67e13b"},
-    {"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789", "d174ab98d277d9f5a5611c2c9f419d9f"},
-    {"12345678901234567890123456789012345678901234567890123456789012345678901234567890",
-     "57edf4a22be3c955ac49da2e2107b67a"}};
-
 class fixture {
 public:
-    md5::construction::type a;
+    accumulator_set<md5> acc;
+    virtual ~fixture() {
+    }
 };
 
-BOOST_AUTO_TEST_SUITE(md5_test_suite)
-
-BOOST_DATA_TEST_CASE(md5_range_hash, boost::unit_test::data::make(string_data), array_element) {
-    std::string out = hash<md5>(array_element.first);
 
-    BOOST_CHECK_EQUAL(out, array_element.second);
-}
+BOOST_AUTO_TEST_SUITE(md5_test_suite)
 
 BOOST_FIXTURE_TEST_CASE(md5_accumulator1, fixture) {
-    md5::construction::type::digest_type s = a.end_message();
-
-#ifdef CRYPTO3_HASH_SHOW_PROGRESS
-    std::printf("%s\n", std::to_string(s));
-#endif
-
-    BOOST_CHECK_EQUAL("0123456789abcdeffedcba9876543210", std::to_string(s));
-    a.reset();
-}
-
-BOOST_FIXTURE_TEST_CASE(md5_accumulator2, fixture) {
-    // 0-length input: echo -n | md4sum
+    // 0-length input: echo -n | md5sum
 
     // A single 1 bit after the (empty) message,
     // then pad with 0s,
-    // then add the length, which is also 0.
-    // Remember that MD5 is little-octet, big-bit endian
-    md5::construction::type::block_type m = {{0x00000080u}};
-    a.process_block(m);
-    md5::construction::type::digest_type s = a.end_message();
+    // then add the length, which is also 0
+    md5::construction::type::block_type m = {{0x80000000u}};
+    acc(m);
+
+    md5::digest_type s = extract::hash<md5>(acc);
 
 #ifdef CRYPTO3_HASH_SHOW_PROGRESS
-    std::printf("%s\n", std::to_string(s));
+    std::printf("%s\n", std::to_string(s).data());
 #endif
 
-    BOOST_CHECK_EQUAL("d41d8cd98f00b204e9800998ecf8427e", std::to_string(s));
-
-    a.reset();
+    BOOST_CHECK_EQUAL("d41d8cd98f00b204e9800998ecf8427e", std::to_string(s).data());
 }
 
-BOOST_FIXTURE_TEST_CASE(md5_accumulator3, fixture) {
-    // echo -n "abc" | md4sum
+BOOST_FIXTURE_TEST_CASE(md5_accumulator2, fixture) {
+    // echo -n "abc" | md5sum
     md5::construction::type::block_type m = {{}};
-    m[0] = 0x80636261;
-    // little-octet, big-bit endian also means the size isn't in the last word
-    m[14] = 0x00000018;
-    a.process_block(m);
-    md5::construction::type::digest_type s = a.end_message();
+    m[0] = 0x00636261;
+    acc(m, nil::crypto3::accumulators::bits = 24);
```

Comment:
I wonder if we could use accumulator without explicit bits mentioning?

Reply:
At the moment we can't. Unless we hash a full block or word. Otherwise, we need any information about the actual size of the message, and we don't have it inside the block.